### PR TITLE
when process dtls packet, only try send data back if handshake is not…

### PR DESCRIPTION
**Pull request by chehefen**
… done. In dtls timer callback, do handle dtls timeout.

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/source/Dtls/Dtls.c [hunk 1]
```diff
@@ -34,17 +34,37 @@ STATUS dtlsTransmissionTimerCallback(UINT32 timerID, UINT64 currentTime, UINT64
...
 
+    MUTEX_LOCK(pDtlsSession->sslLock);
+    locked = TRUE;
+
     CHK(!SSL_is_init_finished(pDtlsSession->pSsl), STATUS_TIMER_QUEUE_STOP_SCHEDULING);
 
-    // do SSL_do_handshake to drive ssl state machine
-    SSL_do_handshake(pDtlsSession->pSsl);
-    CHK_STATUS(dtlsCheckOutgoingDataBuffer(pDtlsSession));
+    if (DTLSv1_get_timeout(pDtlsSession->pSsl, &timeout) == 0) {
```
Comment (Sean-Der) [inline src/source/Dtls/Dtls.c]: Possible to put `DTLSv1_get_timeout` directly inside `CHK` call?

File: src/source/Dtls/Dtls.c [hunk 2]
```diff
@@ -342,13 +362,21 @@ STATUS dtlsSessionProcessPacket(PDtlsSession pDtlsSession, PBYTE pData, PINT32 p
...
+    } else {
+        // if dtls handshake is done, and SSL_read did not fail, then sslRet and number of sctp bytes read
+        dataLen = sslRet < 0 ? 0 : sslRet;
+    }
 
 CleanUp:
 
     CHK_LOG_ERR(retStatus);
 
+    if (pDataLen != NULL) {
```
Comment (Sean-Der) [inline src/source/Dtls/Dtls.c]: I would drop `pDataLen`, we have a `CHK` call when we enter the function

File: src/source/Dtls/Dtls.c [hunk 3]
```diff
@@ -342,13 +362,21 @@ STATUS dtlsSessionProcessPacket(PDtlsSession pDtlsSession, PBYTE pData, PINT32 p
         LOG_OPENSSL_ERROR("SSL_read");
     }
 
-    *pDataLen = sslRet;
-    CHK_STATUS(dtlsCheckOutgoingDataBuffer(pDtlsSession));
+    if (!SSL_is_init_finished(pDtlsSession->pSsl)) {
+        CHK_STATUS(dtlsCheckOutgoingDataBuffer(pDtlsSession));
+    } else {
+        // if dtls handshake is done, and SSL_read did not fail, then sslRet and number of sctp bytes read
+        dataLen = sslRet < 0 ? 0 : sslRet;
```
Comment (Sean-Der) [inline src/source/Dtls/Dtls.c]: I would just set `*pDataLen` directly, I don't believe the indirection is needed.

File: src/source/Dtls/Dtls.c (hunk 2 above)
Comment (chehefen) [reply src/source/Dtls/Dtls.c]: fixed

File: src/source/Dtls/Dtls.c (hunk 1 above)
Comment (chehefen) [reply src/source/Dtls/Dtls.c]: fix

File: .gitignore [hunk 4]
```diff
@@ -11,3 +11,4 @@ open-source/libusrsctp
 open-source/libwebsockets
 open-source/local
 outputs
+open-source/amazon-kinesis-video-streams-producer-sdk-cpp
```
Comment (MushMal) [inline .gitignore]: Are we sure we need to add this to ignore? I mean, there is a way to modify/send CR for the linked project but perhaps we don't want this now

File: src/source/Dtls/Dtls.c [hunk 5]
```diff
@@ -44,11 +44,8 @@ STATUS dtlsTransmissionTimerCallback(UINT32 timerID, UINT64 currentTime, UINT64
     locked = TRUE;
 
     CHK(!SSL_is_init_finished(pDtlsSession->pSsl), STATUS_TIMER_QUEUE_STOP_SCHEDULING);
-
-    if (DTLSv1_get_timeout(pDtlsSession->pSsl, &timeout) == 0) {
-        // failed to get timeout. try again on next iteration
-        CHK(FALSE, retStatus);
-    }
+    // if not timeout yet. try again on next iteration
+    CHK(DTLSv1_get_timeout(pDtlsSession->pSsl, &timeout) == 1, retStatus);
```
Comment (MushMal) [inline src/source/Dtls/Dtls.c]: Not sure about checking part. The documentation doesn't say anything about the successful return being 1

https://boringssl.googlesource.com/boringssl/+/63fa118f3a3d065adb6ca17227bb9e407ffadccb/include/openssl/ssl.h

Perhaps, this should be CHK(DTLSv1_get_timeout(pDtlsSession->pSsl, &timeout) != 0, retStatus);

File: .gitignore (hunk 4 above)
Comment (Sean-Der) [reply .gitignore]: I don't believe we want this, @chehefen was something happening that you needed this? Maybe there is something we need to fix to make this not needed.

**Pull request closed without merging**